src/main.cpp

```diff
@@ -3838,11 +3841,33 @@ bool ReceivedBlockTransactions(
         deque<CBlockIndex*> queue;
         queue.push_back(pindexNew);
 
+        CCoinsViewCache view(pcoinsTip);
+
         // Recursively process any descendant blocks that now may be eligible to be connected.
         while (!queue.empty()) {
             CBlockIndex *pindex = queue.front();
             queue.pop_front();
             pindex->nChainTx = (pindex->pprev ? pindex->pprev->nChainTx : 0) + pindex->nTx;
+
+            // Given that all parents are BLOCK_VALID_TRANSACTIONS, we now should be
+            // able to calculate this block's net transparent value, assuming the
+            // block is correct.
+            CAmount transparentValue = 0;
+            for (auto& tx : block.vtx) {
+                for (auto& out : tx.vout) {
+                    transparentValue += out.nValue;
+                }
+                if (view.HaveInputs(tx)) {
```

Comment:
There is a problem with this that it was happening before moving the block to this spot so it is independent.

Found with the `ReceivedBlockTransactions` gtest:

```
...
[----------] 3 tests from Validation
[ RUN      ] Validation.ContextualCheckInputsPassesWithCoinbase
[       OK ] Validation.ContextualCheckInputsPassesWithCoinbase (0 ms)
[ RUN      ] Validation.ContextualCheckInputsDetectsOldBranchId
[       OK ] Validation.ContextualCheckInputsDetectsOldBranchId (1 ms)
[ RUN      ] Validation.ReceivedBlockTransactions

Thread 1 "zcash-gtest" received signal SIGSEGV, Segmentation fault.
0x0000555555ed0608 in CCoinsViewCache::FetchCoins (this=this@entry=0x7fffffffcaf8, txid=...) at coins.cpp:122
122	    if (!base->GetCoins(txid, tmp))
(gdb) bt
#0  0x0000555555ed0608 in CCoinsViewCache::FetchCoins (this=this@entry=0x7fffffffcaf8, txid=...) at coins.cpp:122
#1  0x0000555555ed4ad0 in CCoinsViewCache::AccessCoins (this=0x7fffffffcaf8, txid=...) at coins.cpp:689
#2  CCoinsViewCache::HaveInputs (this=0x7fffffffcaf8, tx=...) at coins.cpp:1002
#3  0x0000555555c14a57 in ReceivedBlockTransactions (block=..., state=..., chainparams=..., pindexNew=<optimized out>, pindexNew@entry=0x7fffffffcef8, pos=...) at main.cpp:3860
#4  0x0000555555ae2448 in Validation_ReceivedBlockTransactions_Test::TestBody (this=<optimized out>) at gtest/test_validation.cpp:270
#5  0x0000555555b83d24 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
#6  0x0000555555b66273 in testing::Test::Run() ()
#7  0x0000555555b67120 in testing::TestInfo::Run() ()
#8  0x0000555555b678f7 in testing::TestCase::Run() ()
#9  0x0000555555b73f27 in testing::internal::UnitTestImpl::RunAllTests() ()
#10 0x0000555555b84844 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) ()
#11 0x0000555555b739dc in testing::UnitTest::Run() ()
#12 0x0000555555996931 in RUN_ALL_TESTS () at /home/oxarbitrage/zcash/issue4294_2/zcash/depends/x86_64-pc-linux-gnu/share/../include/gtest/gtest.h:2341
#13 main (argc=<optimized out>, argv=0x7fffffffde18) at gtest/main.cpp:48
(gdb) 
```
```

I was not able to find the root cause of it yet.

Reply:
The reason for this bug is line 4565; it accesses the global `pcoinsTip` which is not set up properly by the test that's failing here. I think perhaps instead we should pass the coins view in as an argument.